File: scripts/bridge_test_server.py
"""
Minimal Flask server to exercise BR_Bridge.tdl (v0).
Run: python scripts/bridge_test_server.py
Then trigger Gateway of Tally > Bridge Test (Alt+B) in TallyPrime.
"""

from __future__ import annotations

import json
from http.server import BaseHTTPRequestHandler, HTTPServer
from urllib.parse import parse_qs, urlparse


HOST = "127.0.0.1"
PORT = 8000


class Handler(BaseHTTPRequestHandler):
    def log_message(self, fmt: str, *args) -> None:
        print(f"[bridge] {self.address_string()} - {fmt % args}")

    def do_POST(self) -> None:
        if self.path != "/tally/bridge":
            self.send_error(404, "Not Found")
            return

        length = int(self.headers.get("Content-Length", "0") or 0)
        raw = self.rfile.read(length).decode("utf-8", errors="replace")
        try:
            body = json.loads(raw) if raw else {}
        except json.JSONDecodeError:
            body = {"_parse_error": raw[:500]}

        print("[bridge] headers:", dict(self.headers))
        print("[bridge] body:", body)

        parsed = urlparse(self.path)
        qs = parse_qs(parsed.query)
        mode = (qs.get("mode") or [""])[0]

        if mode == "import":
            xml = (
                "<ENVELOPE>"
                "<HEADER><VERSION>1</VERSION><TALLYREQUEST>Import</TALLYREQUEST>"
                "<TYPE>Data</TYPE><ID>All Masters</ID></HEADER>"
                "<BODY><DATA></DATA></BODY>"
                "</ENVELOPE>"
            )
            out = {"cmd": "import_xml", "xml": xml}
        else:
            out = {"cmd": "show_message", "message": "Bridge connected"}

        payload = json.dumps(out).encode("utf-8")
        self.send_response(200)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)


def main() -> None:
    httpd = HTTPServer((HOST, PORT), Handler)
    print(f"Bridge mock listening on http://{HOST}:{PORT}/tally/bridge")
    httpd.serve_forever()


if __name__ == "__main__":
    main()

File: scripts/bridge-test-server.js
/**
 * Mock bridge API for BR_Bridge.tdl
 * Run: node scripts/bridge-test-server.js
 * Default: show_message. For import_xml test, set URL in TDL to ...?mode=import
 */

const http = require("http");

const HOST = "127.0.0.1";
const PORT = 8000;
const PATH = "/tally/bridge";
const EXPECTED_AUTH = "Bearer dev-token";

function sendJson(res, status, obj) {
  const body = JSON.stringify(obj);
  res.writeHead(status, {
    "Content-Type": "application/json; charset=utf-8",
    "Content-Length": Buffer.byteLength(body),
  });
  res.end(body);
}

const server = http.createServer((req, res) => {
  if (req.method !== "POST" || req.url.split("?")[0] !== PATH) {
    res.writeHead(404);
    res.end();
    return;
  }

  const auth = req.headers.authorization || "";
  if (auth !== EXPECTED_AUTH) {
    console.warn("[bridge] bad Authorization:", auth || "(missing)");
    sendJson(res, 401, { error: "expected Authorization: Bearer dev-token" });
    return;
  }

  let raw = "";
  req.on("data", (c) => {
    raw += c;
  });
  req.on("end", () => {
    let body = {};
    try {
      body = raw ? JSON.parse(raw) : {};
    } catch {
      body = { _raw: raw };
    }
    console.log("[bridge] POST", new Date().toISOString());
    console.log("[bridge] body:", JSON.stringify(body, null, 2));

    const mode = new URL(req.url, "http://127.0.0.1").searchParams.get("mode");
    let out;
    if (mode === "import") {
      out = {
        cmd: "import_xml",
        xml:
          "<ENVELOPE><HEADER><VERSION>1</VERSION><TALLYREQUEST>Import</TALLYREQUEST>" +
          "<TYPE>Data</TYPE><ID>All Masters</ID></HEADER><BODY><DATA></DATA></BODY></ENVELOPE>",
      };
    } else {
      out = { cmd: "show_message", message: "Bridge connected (mock server)" };
    }
    sendJson(res, 200, out);
    console.log("[bridge] response:", out);
  });
});

server.listen(PORT, HOST, () => {
  console.log(`Bridge mock listening on http://${HOST}:${PORT}${PATH}`);
  console.log("In Tally: open a company, Gateway of Tally -> Bridge Test (Alt+B)");
});
